### PR TITLE
implement drain for stocking input hatch

### PR DESCRIPTION
**Pull request by Glease**
This is primarily used by depleteInput() from controller. only drain from internal source (i.e. ForgeDirection.UNKNOWN) is allowed.

### CODE REVIEW

File: src/main/java/gregtech/common/tileentities/machines/GT_MetaTileEntity_Hatch_Input_ME.java [hunk 1]
```diff
@@ -207,6 +207,21 @@ public FluidStack[] getStoredFluids() {
         return shadowStoredFluids;
     }
 
+    @Override
+    public FluidStack drain(ForgeDirection side, FluidStack aFluid, boolean doDrain) {
+        if (false) {
```
Comment (miozune) [inline src/main/java/gregtech/common/tileentities/machines/GT_MetaTileEntity_Hatch_Input_ME.java:212]: leftover

File: src/main/java/gregtech/common/tileentities/machines/GT_MetaTileEntity_Hatch_Input_ME.java [hunk 2]
```diff
@@ -207,6 +207,21 @@ public FluidStack[] getStoredFluids() {
...
+            return super.drain(side, aFluid, doDrain);
+        }
+        if (side != ForgeDirection.UNKNOWN) return null;
```
Comment (miozune) [inline src/main/java/gregtech/common/tileentities/machines/GT_MetaTileEntity_Hatch_Input_ME.java:215]: maybe leave comment on why this method is needed and why it forbids sided extraction?

File: src/main/java/gregtech/common/tileentities/machines/multi/GT_MetaTileEntity_HeatExchanger.java [hunk 3]
```diff
@@ -291,6 +292,7 @@ public boolean onRunningTick(ItemStack aStack) {
                         explodeMultiblock(); // Generate crater
                     }
                 }
+                endRecipeProcessing();
```
Comment (miozune) [inline src/main/java/gregtech/common/tileentities/machines/multi/GT_MetaTileEntity_HeatExchanger.java:295]: I think these are not needed? I made `checkRecipe()` final to make sure every multi calls `startRecipeProcessing` and `endRecipeProcessing`

File: src/main/java/gregtech/common/tileentities/machines/multi/GT_MetaTileEntity_LargeBoiler.java [hunk 4]
```diff
@@ -386,6 +387,7 @@ public boolean onRunningTick(ItemStack aStack) {
                         explodeMultiblock();
                     }
                 }
+                endRecipeProcessing();
```
Comment (miozune) [inline src/main/java/gregtech/common/tileentities/machines/multi/GT_MetaTileEntity_LargeBoiler.java:390]: Same here

File: src/main/java/gregtech/common/tileentities/machines/multi/GT_MetaTileEntity_HeatExchanger.java (hunk 3 above)
Comment (Glease) [reply src/main/java/gregtech/common/tileentities/machines/multi/GT_MetaTileEntity_HeatExchanger.java:295]: This is onRunningTick, not checkRecipe()

File: src/main/java/gregtech/common/tileentities/machines/multi/GT_MetaTileEntity_LargeBoiler.java (hunk 4 above)
Comment (Glease) [reply src/main/java/gregtech/common/tileentities/machines/multi/GT_MetaTileEntity_LargeBoiler.java:390]: This is onRunningTick, not checkRecipe()

File: src/main/java/gregtech/common/tileentities/machines/GT_MetaTileEntity_Hatch_Input_ME.java (hunk 2 above)
Comment (Glease) [reply src/main/java/gregtech/common/tileentities/machines/GT_MetaTileEntity_Hatch_Input_ME.java:214]: added

File: src/main/java/gregtech/common/tileentities/machines/multi/GT_MetaTileEntity_HeatExchanger.java (hunk 3 above)
Comment (miozune) [reply src/main/java/gregtech/common/tileentities/machines/multi/GT_MetaTileEntity_HeatExchanger.java:295]: Oops

File: src/main/java/gregtech/common/tileentities/machines/GT_MetaTileEntity_Hatch_Input_ME.java [hunk 5]
```diff
@@ -207,6 +207,20 @@ public FluidStack[] getStoredFluids() {
         return shadowStoredFluids;
     }
 
+    @Override
+    public FluidStack drain(ForgeDirection side, FluidStack aFluid, boolean doDrain) {
+        // this is an ME output hatch. allowing draining via logistics would be very wrong (and against
```
Comment (miozune) [inline src/main/java/gregtech/common/tileentities/machines/GT_MetaTileEntity_Hatch_Input_ME.java:212]: _input_ hatch

**Review by miozune — APPROVED** [review]
Thanks

**Pull request merged**